crates/sargon/src/hierarchical_deterministic/derivation/derivation_path.rs

```diff
@@ -1,88 +1,191 @@
+            )]
+            pub enum $union_name {
+                $(
+                    #[display("{value}")]
+                    #[debug("{:?}", value)]
+                    $variant_name { value: $variant_type },
+                )+
+            }
 
-impl TryFrom<&HDPath> for DerivationPath {
-    type Error = CommonError;
+            impl $union_name {
+                $(
+                    pub fn [< $variant_name:snake >](path: $variant_type) -> Self {
+                        Self::$variant_name { value: path }
+                    }
+                )+
+            }
 
-    fn try_from(value: &HDPath) -> Result<Self> {
-        if let Ok(bip44) = BIP44LikePath::try_from(value) {
-            return Ok(bip44.into());
-        };
-        CAP26Path::try_from(value).map(|p| p.derivation_path())
-    }
-}
+            impl FromStr for $union_name {
+                type Err = CommonError;
+                fn from_str(s: &str) -> Result<Self> {
+                    Self::from_bip32_string(s)
+                }
+            }
 
-impl FromStr for DerivationPath {
-    type Err = CommonError;
-    fn from_str(s: &str) -> Result<Self> {
-        HDPath::from_str(s).and_then(|p| Self::try_from(&p))
-    }
-}
+            impl $union_name {
 
-impl<'de> serde::Deserialize<'de> for DerivationPath {
-    /// Tries to deserializes a JSON string as a bech32 address into an `AccountAddress`.
-    #[cfg(not(tarpaulin_include))] // false negative
-    fn deserialize<D: Deserializer<'de>>(
-        d: D,
-    ) -> Result<DerivationPath, D::Error> {
-        #[derive(Deserialize)]
-        pub struct Inner {
-            pub scheme: DerivationPathScheme,
-            pub path: serde_json::Value,
-        }
+                pub fn to_hd_path(&self) -> HDPath {
+                    match self {
+                        $(
+                            Self::$variant_name { value } => value.to_hd_path(),
+                        )+
+                    }
+                }
+            }
+            impl From<$union_name> for HDPath {
+                fn from(value: $union_name) -> Self {
+                    value.to_hd_path()
+                }
+            }
 
-        let inner = Inner::deserialize(d)?;
+            impl ToBIP32Str for $union_name {
+                fn to_bip32_string(&self) -> String {
+                    self.to_hd_path().to_bip32_string()
+                }
+                fn to_bip32_string_debug(&self) -> String {
+                    self.to_hd_path().to_bip32_string_debug()
+                }
+            }
+
+            impl FromBIP32Str for $union_name {
+                fn from_bip32_string(s: impl AsRef<str>) -> Result<Self> {
+                    let s = s.as_ref();
+                    Result::<Self>::Err(CommonError::InvalidBIP32Path { bad_value: s.to_owned() })
+                    $(
+                        .or($variant_type::from_bip32_string(s).map(Self::[< $variant_name:snake >]))
+                    )+
```

Comment:
correct me if I am wrong, but in Rust you can return
```Rust
Err(failure_path)
 .or(happy_path)
```
and it will try to perform the `happy_path` even though the `failure_path` is set first?

Reply:
yes that is what I'm doing. Sorry what is the review? :)

Reply:
wasn't a review, just a Rust question since I am surprised this works 🙂 
I was thinking of the `.or()` in Rust as a `??` in Swift, meaning I would have though this would always return an `Err`.

Reply:
looks like there is an error in the documentation of `or` https://doc.rust-lang.org/std/result/enum.Result.html#method.or

it returns the first `Ok`